cargo-watch, stdenv: resolve aarch64-darwin objc redefinition errors

tjni opened the pull request:
###### Description of changes

Resolve compilation failures related to redefining symbols on aarch64-darwin that have broken builds of `cargo-watch`, `starship`, and others.

This includes a revised version of https://github.com/NixOS/nixpkgs/pull/161561 which no longer breaks `xcbuild` and will resolve #163052.

As a whole, these two commits are extracted from a part of the excellent work done in #189977 and resolves #189687 (at least, the errors that result from it) and #146349.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

stephank committed:
darwin.apple_sdk_11_0: use stdenv objc4

Stdenv on aarch64-darwin pulls in (bootstrap-stage4) objc4, unlike
x86_64. However derivations that otherwise depend on objc4 would use a
a different objc4 derivation on top of the final stdenv.

Because this library defines an LLVM module, having multiple instances
of it in the import path will interfere with builds.

stephank committed:
IOSurface: remove xpc dependency

toonn commented on pkgs/os-specific/darwin/apple-sdk-11.0/frameworks.nix:
```diff
@@ -1,4 +1,4 @@
-{ frameworks, libs, libobjc, Libsystem, libnetwork }: with frameworks; with libs;
```
@stephank, what exactly do we achieve by dropping `Libsystem` here? I.e., why can't we pass it explicitly to `IOSurface`?

stephank replied:
This is just an unused parameter, and actually something @tjni added to the commit, I think. I didn't notice it.

As for why not depend on libsystem (instead of xpc), I didn't see any use for something depending on libsystem at all? Isn't it already part of stdenv? I haven't seen any problems leaving it out, but also have not been building much beyond swift and mpv.

toonn replied:
Ok, same question to @tjni then : )

Should we not be explicit in our Libsystem dependencies? I know stdenv brings it along implicitly but if being explicit doesn't hurt it makes things a lot easier to understand.

(Also, if you changed the commits, you as author seems appropriate, maybe add `Co-authored-by:` if they're still based on stephank's work though.)

tjni replied:
I also prefer the explicitness, but we actually cannot add the `Libsystem` dependency, because `cc-wrapper` adds the C++ standard library after these dependencies, and both `libc++` and `libstdc++` require `libc` to come after them. I wrote a little bit more about this in https://github.com/NixOS/nixpkgs/pull/185569#issuecomment-1234959249.

Good idea about keeping the authorship intact. I'll try to do this and move the `Libsystem` parameter cleanup into its own commit.

toonn replied:
Maybe `libstdc++` and `libc++` should be prefixed rather than suffixed? If that doesn't mess anything up but allows us to be explicit about things that'd be a net benefit I think.

If you think the interactions of this would be too hard to foresee then I'm fine with the current state.

tjni replied:
The am not sure it's the correct thing to do because of the following lookup order:

> The lookup order is as follows:
>
>1. For the quote form of the include directive, the directory of the current file is searched first.
>2. For the quote form of the include directive, the directories specified by -iquote options are searched in left-to-right >order, as they appear on the command line.
>3. Directories specified with -I options are scanned in left-to-right order.
>4. Directories specified with -isystem options are scanned in left-to-right order.
>5. Standard system directories are scanned.
>6. Directories specified with -idirafter options are scanned in left-to-right order.

We add dependency headers using `-isystem`, presumably so that any warnings in dependencies are ignored when compiling the current project. Meanwhile, many a non-nixpkgs build would pick up `lib[std]c++` and `libc` from the standard system directories. We emulate this in nixpkgs by passing `lib[std]c++` and `libc` after other dependencies.

Without knowing any more details, I presume the lookup order was selected intentionally so that we should keep to the current behavior in `cc-wrapper`.

toonn replied:
I feel like this is valuable reverse engineering, even if it's in part speculative. Do you think adding this as a comment in an appropriate place would allow the next person looking into it to have an easier time or refute the reasoning and be motivated to provide a better explanation?

stephank replied:
I believe `-idirafter` also interacts with some uses of `#include_next`, where a header 'shadows' another for some reason. I think I saw this somewhere in debugging an error in the Swift build once, but don't remember what it was.

Does anyone know why IOSurface is special at all? Why does it need the libsystem dependency?

toonn replied:
I think it needs XPC, but because Libsystem also provides XPC two versions ended up in things. The first attempted fix was to have it depend on Libsystem instead but that still lead to two different versions.

tjni replied:
@toonn I added a commit with a comment including more context to `cc-wrapper`. It is a good suggestion. Thank you.

toonn reviewed (commented):
This looks pretty good to me but I want to make sure I understand what's going on.

@tjni, since this work really comes from @stephank, can you set him as author on the commits?

tjni committed:
apple_sdk: clean up unused Libsystem parameter

tjni committed:
cc-wrapper: comment explaining C++ stdlib order

There is context here that I needed when resolving an issue in which
libc was added to NIX_CFLAGS_COMPILE before the C++ stdlib that took
me awhile to understand.

It was suggested to me that this context be included as a comment,
since it is not obvious and could help others in the future.

toonn commented:
Staging moved under me while merging so I did a `pull --rebase` neglecting the fact that rewrites commits : s

So this is merged here, https://github.com/NixOS/nixpkgs/commit/c319d8ae3b7522656f74085281672f61c7112a98 and lesson learned to be more careful next time.
